src/meta/app/src/principal/user_privilege.rs

```diff
@@ -66,6 +66,15 @@ pub enum UserPrivilegeType {
     CreateDataMask = 1 << 16,
     // Privilege to Own a databend object such as database/table.
     Ownership = 1 << 17,
+    // Privilege to use external stage
+    UsageExternalStage = 1 << 18,
```

Comment:
what is the meaning of Usage here? R+W?
why is it diff from Internel stage?

Reply:
Same with sf.

UsageUDF: if the user create/drop/alter UDF, need this priv.

UsageExternalStage: if user operator external stage, need this priv.

WriteInternalStage: copy table into stage; presign upload stage

ReadInternalStage: copy stage into table; presign download stage

Reply:
> what is the meaning of Usage here? R+W?
> 
> why is it diff from Internel stage?

https://docs.snowflake.com/en/user-guide/security-access-control-privileges#stage-privileges

Reply:
so why sf do it this way？

Reply:
it seems to me there is  no difference  between InternalStage and ExternalStage when read/write.
except presign?

Reply:
@TCeason 
Databend is a point that difference with snowflake:
snowflake can't allow user to write micro-partition data to externalstage but Databend can.

Reply:
> UsageExternalStage: if user operator external stage, need this priv.
> 
> WriteInternalStage: copy table into stage; presign upload stage
> 
> ReadInternalStage: copy stage into table; presign download stage

So in Databend, we need this four privileges about stage:

```
WriteExternalStage

ReadExternalStage

WriteInternalStage

ReadInternalStage
```

Reply:
why not  just

```
WriteStage
ReadStage
```

Reply:
> @TCeason Databend is a point that difference with snowflake: snowflake can't allow user to write micro-partition data to externalstage but Databend can.

Sorry for confusing, this is not related to stage but to external table, please ignored.

Reply:
guess reasons:

1. may be for some exteranl stage，some operation involve both R/W？ or can not distinguish？but user only access them though SQL COPY and select？
2.  not need to be strict for external?

Reply:
if the logic is:

 some reason
=> external stage not distingush R/W
=> external stage can not use R/W privialge
=> add an USAGE for external stage

Reply:
@flaneur2020 

if UserPrivilegeType need to used with a object

why we can not only use CREAET instead of CREATE，CREATE_USER,CREATE_ROLE?